opm/simulators/linalg/bda/BILU0.cpp

```diff
@@ -0,0 +1,381 @@
+            std::ostringstream out;
+            out << "BILU0 reorder matrix: " << t2 - t1 << " s";
+            OpmLog::info(out.str());
+        }
+
+        // TODO: remove this copy by replacing inplace ilu decomp by out-of-place ilu decomp
+        if (verbosity >= 3){
+            t1 = second();
+        }
+        memcpy(LUMat->nnzValues, rMat->nnzValues, sizeof(double) * bs * bs * rMat->nnzbs);
+        if (verbosity >= 3){
+            t2 = second();
+            std::ostringstream out;
+            out << "BILU0 memcpy: " << t2 - t1 << " s";
+            OpmLog::info(out.str());
+        }
+
+        int i, j, ij, ik, jk;
+        int iRowStart, iRowEnd, jRowEnd;
+        double pivot[bs * bs];
+
+        int LSize = 0;
+
+        if (verbosity >= 3){
+            t1 = second();
+        }
+        // go through all rows
+        for (i = 0; i < LUMat->Nb; i++) {
+            iRowStart = LUMat->rowPointers[i];
+            iRowEnd = LUMat->rowPointers[i + 1];
+
+            // go through all elements of the row
+            for (ij = iRowStart; ij < iRowEnd; ij++) {
+                j = LUMat->colIndices[ij];
+                // if the element is the diagonal, store the index and go to next row
+                if (j == i) {
+                    diagIndex[i] = ij;
+                    break;
+                }
+                // if an element beyond the diagonal is reach, no diagonal was found
+                // throw an error now. TODO: perform reordering earlier to prevent this
+                if (j > i) {
+                    std::ostringstream out;
+                    out << "BILU0 Error could not find diagonal value in row: " << i;
+                    OpmLog::error(out.str());
+                    return false;
+                }
+
+                LSize++;
+                // calculate the pivot of this row
+                blockMult<bs>(LUMat->nnzValues + ij * bs * bs, invDiagVals + j * bs * bs, &pivot[0]);
+
+                memcpy(LUMat->nnzValues + ij * bs * bs, &pivot[0], sizeof(double) * block_size * block_size);
+
+                jRowEnd = LUMat->rowPointers[j + 1];
+                jk = diagIndex[j] + 1;
+                ik = ij + 1;
+                // substract that row scaled by the pivot from this row.
+                while (ik < iRowEnd && jk < jRowEnd) {
+                    if (LUMat->colIndices[ik] == LUMat->colIndices[jk]) {
+                        blockMultSub<bs>(LUMat->nnzValues + ik * bs * bs, pivot, LUMat->nnzValues + jk * bs * bs);
+                        ik++;
+                        jk++;
+                    } else {
+                        if (LUMat->colIndices[ik] < LUMat->colIndices[jk])
+                        { ik++; }
+                        else
+                        { jk++; }
+                    }
+                }
+            }
+            // store the inverse in the diagonal!
+            blockInvert3x3(LUMat->nnzValues + ij * bs * bs, invDiagVals + i * bs * bs);
```

Comment:
I forgot to template this, I'd rather copy the 2x2 and 4x4 codes from `MatrixBlock.hpp` then construct a FieldMatrix here.

Reply:
Not sure this applies here, but in general the maintenance burden increases when copying code as bug fixes are usually not copied.
My idea was to put the inversion algorithms (in MatrixBlock.hpp) into a separate function that just gets C-arrays and works on them. These functions can then be called for FieldMatrix, MatrixBlock, and BlockedMartix. Hence we also reduce code copying in the original header MatrixBlock.hpp.

Reply:
I'm trying to get these functions to MatrixBlock.hpp, but I have a problem with the templates. I need to specify inversions for blocksizes 1 to 4. And the existing algorithms allow for both double and floats (and maybe other types). It would probably work if only doubles are needed.
I have reduced the problem [here](https://ideone.com/vkVcN7).

Reply:
there are probably more elegant solutions, but one (old school) is
```
#include <iostream>
using namespace std;

template<int BS>
struct Inverter
{
    template<typename K>
    void operator()(const K *matrix, K *inverse)
    {
        throw std::logic_error("Not implemented");
    }
};

template<>
struct Inverter<1>
{
    template<typename K>
    void operator()(const K *matrix, K *inverse)
    {
    }
};

int main() {
	double *a = new double[4];
	double *b = new double[4];
	Inverter<1> invert;
        invert(a, b);
	
	return 0;
}
```

Reply:
Implemented in 833ea8a